rosette-scala/src/main/scala/coop/rchain/rosette/Location.scala

```diff
@@ -149,90 +178,222 @@ object Location {
-        v.setAddr(indirect, level, offset, value)
+      case LocationGT(genericType) =>
+        genericType match {
+          case LTLexVariable(indirect, level, offset) =>
+            v.setLex(indirect, level, offset, value)
 
-      case LTGlobalVariable(offset) =>
-        globalEnv.setLex(1, 0, offset, value)
+          case LTAddrVariable(indirect, level, offset) =>
+            v.setAddr(indirect, level, offset, value)
 
-      case LTBitField(indirect, level, offset, spanSize, sign) =>
-        v.setField(indirect, level, offset, spanSize, fixVal(value))
+          case LTGlobalVariable(offset) =>
+            globalEnv.setLex(1, 0, offset, value)
 
-      case LTBitField00(offset, spanSize, sign) =>
-        v.setField(0, 0, offset, spanSize, fixVal(value))
+          case LTBitField(indirect, level, offset, spanSize, sign) =>
+            v.setField(indirect, level, offset, spanSize, fixVal(value))
 
-      case _ => {
-        suicide("setValWrt")
+          case LTBitField00(offset, spanSize, sign) =>
+            v.setField(0, 0, offset, spanSize, fixVal(value))
+
+          case _ => {
+            suicide(s"Location.setValWrt: $loc")
+            null
+          }
+        }
+    }
+
+  def adjustLevel(loc: Location, adjustment: Int): Location =
+    loc match {
+      case LocationAtom(_) => {
+        suicide(s"Location.adjustLevel: $loc")
         null
       }
+
+      case LocationGT(genericType) =>
+        LocationGT(genericType match {
+          case LTLexVariable(indirect, level, offset) =>
+            LTLexVariable(indirect, level + adjustment, offset)
+
+          case LTAddrVariable(indirect, level, offset) =>
+            LTAddrVariable(indirect, level + adjustment, offset)
+
+          case LTBitField(indirect, level, offset, spanSize, sign) =>
+            LTBitField(indirect, level + adjustment, offset, spanSize, sign)
+
+          case _ => {
+            suicide(s"Location.adjustLevel: $loc")
+            null
+          }
+        })
+    }
+
+  def CtxtReg(n: Int): Location = {
+    if (n >= NumberOfCtxtRegs) {
+      suicide(s"Location.CtxtReg: invalid ctxt register ($n)");
+      null
+    }
+
+    LocationGT(LTCtxtRegister(n))
+  }
+
+  def ArgReg(n: Int): Location = {
+    if (n > MaxArgs) {
+      suicide(s"Location.ArgReg: invalid arg register index ($n)")
+      null
     }
+
+    LocationGT(LTArgRegister(n))
+  }
+
+  def LexVar(level: Int, offset: Int, indirect: Int): Location = {
+    if (level >= (1 << LexLevelSize) || offset >= (1 << LexOffsetSize)) {
+      val offsetStr = if (indirect != 0) s"($offset)" else s"$offset"
+      suicide(
+        s"Location.LexVar: unrepresentable location (lex[$level,$offsetStr])")
+      null
+    }
+    LocationGT(LTLexVariable(if (indirect == 0) 0 else 1, level, offset))
+  }
+
+  def AddrVar(level: Int, offset: Int, indirect: Int): Location = {
+    if (level >= (1 << LexLevelSize) || offset >= (1 << LexOffsetSize)) {
```

Comment:
Should be `AddrLevelSize` and `AddrOffsetSize`.

Reply:
Fixed